ocaml-lsp-server/src/workspace_symbol.ml

```diff
@@ -0,0 +1,167 @@
+open Import
+
+let cmt_sign (cmt_infos : Cmt_format.cmt_infos) =
+  match cmt_infos.cmt_annots with
+  | Cmt_format.Implementation { Typedtree.str_type = sign; _ }
+  | Cmt_format.Interface { Typedtree.sig_type = sign; _ }
+  | Cmt_format.Packed (sign, _) ->
+    sign
+  | _ -> []
+
+let loc_of_sig_item (signature_item : Types.signature_item) =
+  match signature_item with
+  | Types.Sig_value (_, descr, _) -> descr.val_loc
+  | Types.Sig_type (_, descr, _, _) -> descr.type_loc
+  | Types.Sig_typext (_, descr, _, _) -> descr.ext_loc
+  | Types.Sig_module (_, _, descr, _, _) -> descr.md_loc
+  | Types.Sig_modtype (_, descr, _) -> descr.mtd_loc
+  | Types.Sig_class (_, descr, _, _) -> descr.cty_loc
+  | Types.Sig_class_type (_, descr, _, _) -> descr.clty_loc
+
+let id_of_sig_item (signature_item : Types.signature_item) =
+  match signature_item with
+  | Types.Sig_value (id, _, _)
+  | Types.Sig_type (id, _, _, _)
+  | Types.Sig_typext (id, _, _, _)
+  | Types.Sig_module (id, _, _, _, _)
+  | Types.Sig_modtype (id, _, _)
+  | Types.Sig_class (id, _, _, _)
+  | Types.Sig_class_type (id, _, _, _) ->
+    id
+
+let kind_of_sig_item (signature_item : Types.signature_item) : SymbolKind.t =
+  match signature_item with
+  | Types.Sig_value _ -> Function
+  | Types.Sig_type _ -> String
+  (* | Types.Sig_typext (_, _, Types.Text_exception, _) -> Constructor *)
+  | Types.Sig_typext _ -> Constructor
+  | Types.Sig_module _ -> Module
+  | Types.Sig_modtype _ -> Module
+  | Types.Sig_class _ -> Class
+  | Types.Sig_class_type _ -> Class
+
+let symbol_of_signature_item ~sourcepath (signature_item : Types.signature_item)
+    =
+  let id = id_of_sig_item signature_item in
+  let loc = loc_of_sig_item signature_item in
+  let name = Ident.name id in
+  let kind = kind_of_sig_item signature_item in
+  let range = Range.of_loc loc in
+  let uri = sourcepath |> Uri.of_path in
+  let location = Location.create ~uri ~range in
+  SymbolInformation.create ~name ~kind ~location ()
+
+let is_directory dir =
+  try Sys.is_directory dir with
+  | Sys_error _ -> false
+
+type error = Build_dir_not_found of string
+
+let find_build_dir ({ name; uri } : WorkspaceFolder.t) =
+  let build_dir = Filename.concat (Uri.to_path uri) "_build/default" in
+  if is_directory build_dir then
+    Ok build_dir
+  else
+    Error (Build_dir_not_found name)
+
+type cm_file =
+  | Cmt of string
+  | Cmi of string
+
+let string_of_cm cm =
+  match cm with
+  | Cmt f
+  | Cmi f ->
+    f
+
+let symbols_from_cm_file root_uri cm_file =
+  let filename_from_path path = Filename.basename path in
+  let filename = string_of_cm cm_file in
+  let cmt = Cmt_format.read_cmt filename in
```

Comment:
This seems wrong to me - it clearly cannot work for cmi files. Shouldn't it be `Cmi_format.read_cmi` in that case?

Reply:
I switched over to using `cmti` files instead of `cmi`. It appears that `cmti` [is what gets created for `mli` files](https://discord.com/channels/436568060288172042/436568060288172044/824343476900593704)

The advantage is that they can be read with the same `Cmt_format.read` function and thus return the same type.